Sources_v3/StreamChat/Workers/ChatClientUpdater_Tests.swift

```diff
@@ -0,0 +1,523 @@
+            }
+
+            // Simulate error while establishing a connection.
+            client.completeConnectionIdWaiters(connectionId: nil)
+
+            // Wait completion is called.
+            AssertAsync.willBeTrue(connectCompletionCalled)
+
+            if connectionError == nil {
+                // Assert `ClientError.ConnectionNotSuccessful` error is propagated.
+                XCTAssertTrue(connectCompletionError is ClientError.ConnectionNotSuccessfull)
+            } else {
+                // Assert `ClientError.ConnectionNotSuccessful` error with underlaying error is propagated.
+                let clientError = connectCompletionError as! ClientError.ConnectionNotSuccessfull
+                XCTAssertTrue(clientError.underlyingError is ClientError.Unexpected)
+            }
+        }
+    }
+
+    func test_connect_keepsUpdaterAlive() throws {
+        // Create an active client with user session.
+        let client = mockClientWithUserSession()
+
+        // Create an updater.
+        var updater: ChatClientUpdater<ExtraData>? = .init(client: client)
+
+        // Disconnect client from web-socket.
+        updater?.disconnect()
+
+        // Simulate `connect` call and catch the result.
+        updater?.connect()
+
+        // Create weak ref and drop a strong one.
+        weak var weakUpdater = updater
+        updater = nil
+
+        // Assert `connect` keeps updater alive.
+        AssertAsync.staysTrue(weakUpdater != nil)
+
+        // Simulate established connection.
+        client.completeConnectionIdWaiters(connectionId: .unique)
+
+        // Assert updater is deallocated.
+        AssertAsync.willBeNil(weakUpdater)
+    }
+
+    // MARK: Reload User
+
+    func test_reloadUserIfNeeded_currentUser_sameToken_happyPath() throws {
+        // Create an active client with user session.
+        let client = mockClientWithUserSession()
+
+        // Create an updater.
+        let updater = ChatClientUpdater<ExtraData>(client: client)
+
+        // Save current background worker ids.
+        let oldWorkderIDs = client.testBackgroundWorkerIDs
+
+        // Call `reloadUserIfNeeded` without changing a token provider
+        // and synchronously get the result since no job should be done.
+        let error = try await(updater.reloadUserIfNeeded)
+
+        // Assert error is nil.
+        XCTAssertNil(error)
+        // Assert `disconnect` was not called.
+        XCTAssertEqual(client.mockWebSocketClient.disconnect_calledCounter, 0)
+        // Assert `completeTokenWaiters` was not called since token is not changed.
+        XCTAssertFalse(client.completeTokenWaiters_called)
+        // Assert background workers stay the same.
+        XCTAssertEqual(client.testBackgroundWorkerIDs, oldWorkderIDs)
+        // Assert database is not flushed.
+        XCTAssertFalse(client.mockDatabaseContainer.removeAllData_called)
+    }
+
+    func test_reloadUserIfNeeded_happyPaths() throws {
```

Comment:
This is one hell of a test 🤯

Reply:
That's not surprising 😄 Changing a user is indeed the most complex part of the SDK I've ever faced with